### DISCUSSION TITLE
Symbolic vector of functions to apply matrix operations

### DISCUSSION THREAD

**Discussion by jgduenasl** [sympy/sympy#25935] [General]
I am trying to solve an ordinary differential equation system. It consists of three equations for functions  $H, L, C$ on time $t$. These equations are coupled, so if we use matrix notation it can be recast as a single vector differential equation. To do this, I define the column vector $\mathbf{P}^{T} = (H, L, C)$, with $T$ the transpose operation, and a coupling matrix 

$$\mathbf{\Lambda} = 
\left(\begin{array}{ccc}
\mu & -\mu & 0 \\\
-\mu & \mu & -2 \\\
0 & 2 & \mu
\end{array}\right\)
$$ 

with $\mu$ constant, such that the system is

$$
\frac{d}{dt}
\left(\begin{array}{c}
H \\\
L \\\
C 
\end{array}\right\)(t) =
\left(\begin{array}{ccc}
\mu & -\mu & 0 \\\
-\mu & \mu & -2 \\\
0 & 2 & \mu
\end{array}\right\)
\left(\begin{array}{c}
H \\\
L \\\
C 
\end{array}\right\)(t)
$$

If I assume $H, L, C$ just as symbols and define the vector $\mathbf{P}$, I can diagonalize the $\Lambda$ matrix and solve each uncoupled equation at hand or with `dsolve()` method as shown in the [documentation](https://docs.sympy.org/latest/guides/solving/solve-ode.html). The diagonalization is made by
```python
import sympy
from sympy import symbols, simplify, cse, Eq, pprint, shape
from sympy.matrices import Matrix, eye, zeros, ones, diag

from sympy.interactive.printing import init_printing
init_printing(use_unicode=True, wrap_line=False)

t, mu = symbols('t mu', real = True)

H, L, C = symbols('H L C')
P = Matrix([H, L, C]); P

Lambda = Matrix([[mu, -mu, 0], [-mu, mu, -2], [0, 2, mu]]); Lambda
S, J = Lambda.diagonalize()

```
However, I want to solve the system using matrix operations for $\mathbf{P}$ and the methods for differential equation system in matrix notation. I mean, to solve the vector equation 

$$ \frac{d}{dt}\mathbf{P}(t) = \mathbf{\Lambda}\cdot\mathbf{P}(t)$$

for which I need to define $H, L, C$ as function symbols. I am trying to do it as

```python
import sympy
from sympy import symbols, simplify, cse, Eq, pprint, shape
from sympy.matrices import Matrix, eye, zeros, ones, diag

from sympy.interactive.printing import init_printing
init_printing(use_unicode=True, wrap_line=False)

from sympy import Function, Derivative, dsolve

t, mu = symbols('t mu', real=True)
H, L, C = symbols('H L C', cls=Function)

P = Matrix([H, L, C]); P
assert shape(P) == (3,1)

dP = Matrix([Derivative(field, t) for field in P]); dP
assert shape(dP) == (3,1)

coupling_matrix = Matrix([[mu, -mu, 0], [-mu, mu, -2], [0, 2, mu]])
coupling_matrix
shape(coupling_matrix)

# TODO: how should multiply coupling_matrix and P?
diff_eq_sys = Eq(dP -  coupling_matrix * P, zeros(*shape(P)))
diff_eq_sys
```
this code generates the next warning to the definition of a vector of symbolic functions that it was deprecated. 

```vim
7 >>> P = Matrix([H, L, C]); P
  6 <stdin>:1: SymPyDeprecationWarning:
  5
  4 non-Expr objects in a Matrix is deprecated. Matrix represents
  3 a mathematical matrix. To represent a container of non-numeric
  2 entities, Use a list of lists, TableForm, NumPy array, or some
  1 other data structure instead.
41
  1 See https://docs.sympy.org/latest/explanation/active-deprecations.html#deprecated-non-expr-in-matrix
  2 for details.
  3
  4 This has been deprecated since SymPy version 1.9. It
  5 will be removed in a future version of SymPy.
```
It implies that I could not perform the multiplication of $\mathbf{P}$ with the coupling matrix $\mathbf{\Lambda}$ as

```vim
>>> coupling_matrix * P
  9 Traceback (most recent call last):
  8   File "<stdin>", line 1, in <module>
  7   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/core/decorators.py", line 106, in binary_op_wrapper
  6     return func(self, other)
  5   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/matrices/common.py", line 2702, in __mul__
  4     return self.multiply(other)
  3   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/matrices/common.py", line 2729, in multiply
  2     m = self._eval_matrix_mul(other)
  1   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/matrices/repmatrix.py", line 249, in _eval_matrix_mul
190     return classof(self, other)._fromrep(self._rep * other._rep)
  1   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/polys/matrices/domainmatrix.py", line 813, in __mul__
  2     return A.matmul(B)
  3   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/polys/matrices/domainmatrix.py", line 1063, in matmul
  4     return A.from_rep(A.rep.matmul(B.rep))
  5   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/polys/matrices/sdm.py", line 483, in matmul
  6     C = sdm_matmul(A, B, A.domain, m, o)
  7   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/polys/matrices/sdm.py", line 952, in sdm_matmul                                       8     return sdm_matmul_exraw(A, B, K, m, o)                                                                                                      9   File "/opt/anaconda3/lib/python3.10/site-packages/sympy/polys/matrices/sdm.py", line 998, in sdm_matmul_exraw                                10     Ci_list[j].append(Aik * Bkj)                                                                                                               11 TypeError: unsupported operand type(s) for *: 'Symbol' and 'UndefinedFunction'     
```
A possible solution is to define $\mathbf{P}$ as a numpy array and `lambdify()` the matrix $\mathbf{\Lambda}$ to multiply them. It implies to evaluate the lambdified $\mathbf{\Lambda}$ matrix for a specific value for $\mu$ before solving the differential system. 

I wonder if there is a way to define a symbolic vector of functions to apply matrix operations as the multiplication for symbolic matrix $\mathbf{\Lambda}$ ?

**Comment by oscarbenjamin** [comment]
You need to change this:
```python
t, mu = symbols('t mu', real=True)
H, L, C = symbols('H L C', cls=Function)

P = Matrix([H, L, C]);
```
The object `H` here is a function but you need to call the function:
```python
In [7]: H
Out[7]: H

In [8]: H(t)
Out[8]: H(t)
```

**Reply by jgduenasl** [reply]
It works, which means I am losing a conceptual issue between the object $H$ and the calling $H(t)$. I will keep this conversation open until put the solution of the differential system, or the problems found, thxs.

**Comment by oscarbenjamin** [comment]
I'm going to convert this into a discussion because it does not seem to be reporting an issue with sympy.

**Reply by jgduenasl** [reply]
Thanks @oscarbenjamin for your solution. However, I am not understanding the difference between the function object and its calling. When, I type `H = symbols('H', cls=Function)` I am making an instance of [`sympy.core.function.Function(*args)`](https://docs.sympy.org/latest/modules/functions/index.html) class. However, when I call it as `H(t)` I obtained a object `H` as if it was a kind of class. Is it working as a pointer to the location in  memory where the instance `H` is stored? or what is the difference between the calling and the symbol defining a function?
```vim
  5 >>> H = symbols('H', cls=Function)
  4 >>> type(H)
  3 <class 'sympy.core.function.UndefinedFunction'>
  2 >>> type(H(t))
  1 H
```
